### ISSUE TITLE
wasm2c: Are module instances truly thread-safe?

### ISSUE THREAD

**Issue by jonathanmarvens** [WebAssembly/wabt#2372]
Assuming I’ve properly initialized each thread via ‎`wasm_rt_init_thread()`, is it undefined behavior to instantiate a module in thread A and then call into the instance in threads B, C, D, etc.? We can assume only one thread ever calls into the instance at any time.

Based on some quick tests, this behavior appears to be supported, but I’m not sure I trust it—I think I might just be a bit lucky at the moment.

I’m happy to investigate this further on my own, but I thought I’d just ask before going down that rabbit-hole.

Thank you!

**Comment by jonathanmarvens** [comment]
Upon further investigation, the commentary on #2189 and #2332 seem to imply that the module instances are, in fact, thread-safe, but I guess I’m still not sure if the semantics support multiple threads calling into the instance in a well-defined manner (as I described above).

**Comment by keithw** [comment]
Normally when we say code is "thread-safe," we're talking about different threads executing code in parallel. Happy to go down that road, but that doesn't seem to be what you're asking about here. You're just saying, is it okay if a module is instantiated by thread A, and then (after that has provably completed) thread B calls an exported function from that module instance, and then (after that has provably completed) thread C calls an exported function from the same module instance?

Certainly our intention is that this is an easy case and there's no problem. There's no long-lived association between any per-thread state and any particular module instance.

**Comment by jonathanmarvens** [comment]
> If I understand the question, you're just saying, is it okay if a module is instantiated by thread A, and then (after that has provably completed) thread B calls an exported function from that module instance, and then (after that has provably completed) thread C calls an exported function from the same module instance?

Yes, this is exactly what I'm trying to understand. My apologies if I wasn't clear enough.

> Certainly our intention is that this is an easy case and there's no problem. There's no long-lived association between any per-thread state and any particular module instance.

Awesome. In other words, if I encounter any state and data race-related issues with this type of usage, I can assume it's a bug?

**Comment by keithw** [comment]
> Awesome. In other words, if I encounter any state and data race-related issues with this type of usage, I can assume it's a bug?

Yeah, definitely please let us know.

**Comment by jonathanmarvens** [comment]
> Yeah, definitely please let us know.

Will do. Thank you!

(I'll close this issue now, since I got my answer.)
